osu.Game.Rulesets.Osu/Edit/PolygonGenerationPopover.cs

```diff
@@ -136,39 +138,58 @@ private void tryCreatePolygon()
             double length = distanceSnapInput.Current.Value * velocity * timeSpacing;
             float polygonRadius = (float)(length / (2 * Math.Sin(double.Pi / pointInput.Current.Value)));
 
-            editorBeatmap.RemoveRange(insertedCircles);
-            insertedCircles.Clear();
+            int totalPoints = pointInput.Current.Value * repeatCountInput.Current.Value;
+
+            if (insertedCircles.Count > totalPoints)
+            {
+                editorBeatmap.RemoveRange(insertedCircles.GetRange(totalPoints + 1, insertedCircles.Count - totalPoints - 1));
+                insertedCircles.RemoveRange(totalPoints + 1, insertedCircles.Count - totalPoints - 1);
+            }
 
             var selectionHandler = (EditorSelectionHandler)composer.BlueprintContainer.SelectionHandler;
             bool first = true;
 
-            for (int i = 1; i <= pointInput.Current.Value * repeatCountInput.Current.Value; ++i)
+            var newlyAdded = new List<HitCircle>();
+
+            for (int i = 0; i < totalPoints; ++i)
             {
-                float angle = float.DegreesToRadians(offsetAngleInput.Current.Value) + i * (2 * float.Pi / pointInput.Current.Value);
+                float angle = float.DegreesToRadians(offsetAngleInput.Current.Value) + (i + 1) * (2 * float.Pi / pointInput.Current.Value);
                 var position = OsuPlayfield.BASE_SIZE / 2 + new Vector2(polygonRadius * float.Cos(angle), polygonRadius * float.Sin(angle));
 
-                var circle = new HitCircle
-                {
-                    Position = position,
-                    StartTime = startTime,
-                    NewCombo = first && selectionHandler.SelectionNewComboState.Value == TernaryState.True,
-                };
-                // TODO: probably ensure samples also follow current ternary status (not trivial)
-                circle.Samples.Add(circle.CreateHitSampleInfo());
+                bool alreadyAdded = i < insertedCircles.Count;
+
+                var circle = alreadyAdded
+                    ? insertedCircles[i]
+                    : new HitCircle();
+
+                circle.Position = position;
+                circle.StartTime = startTime;
+                circle.NewCombo = first && selectionHandler.SelectionNewComboState.Value == TernaryState.True;
```

Comment:
You can't be changing stuff on hitobjects that already exist in `EditorBeatmap` without calling `EditorBeatmap.Update()` at the end. While yes, stuff like changing start time will automatically do it, stuff like *changing new combo flags* does not. And in fact this behaviour with the combo toggle state is completely broken with this change if the first object is reused.

(Note though that it is half-broken already on `master` if you open the polygon tool without an object selected, because `EditorSelectionHandler` will reset new combo state to off on deselect. You can do

```diff
diff --git a/osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs b/osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs
index dbbf767a7d..dfda93cedd 100644
--- a/osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs
+++ b/osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs
@@ -176,7 +176,8 @@ private void createStateBindables()
         /// </summary>
         protected virtual void UpdateTernaryStates()
         {
-            SelectionNewComboState.Value = GetStateFromSelection(SelectedItems.OfType<IHasComboInformation>(), h => h.NewCombo);
+            if (SelectedItems.Any())
+                SelectionNewComboState.Value = GetStateFromSelection(SelectedItems.OfType<IHasComboInformation>(), h => h.NewCombo);
 
             var samplesInSelection = SelectedItems.SelectMany(enumerateAllSamples).ToArray();
 

```

to change that, and we should probably do that at some point.)

Reply:
Oh, I wasn't aware of needing to call `EditorBeatmap.Update`, I'll keep that in mind for the future. 
I restructured the loop body a bit to clearly split the two cases of adding/updating an object.

The diff above works for preserving the newCombo state in most cases, however when a new object gets added to the beatmap it still resets to false (because of `ComposeBlueprintContainer.hitObjectAdded`). I added an extra bit at the end to restore the previous newcombo state in that case.